src/main/java/org/folio/service/orders/CompositeOrderTotalFieldsPopulateService.java

```diff
@@ -0,0 +1,103 @@
+import org.folio.rest.acq.model.invoice.Invoice;
+import org.folio.rest.acq.model.invoice.InvoiceLine;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.service.finance.transaction.TransactionService;
+import org.folio.service.invoice.InvoiceLineService;
+import org.folio.service.invoice.InvoiceService;
+import org.javamoney.moneta.Money;
+import org.javamoney.moneta.function.MonetaryOperators;
+
+import io.vertx.core.Future;
+
+import static org.folio.orders.utils.HelperUtils.collectResultsOnSuccess;
+
+@Log4j2
+public class CompositeOrderTotalFieldsPopulateService implements CompositeOrderDynamicDataPopulateService {
+
+  private final TransactionService transactionService;
+  private final InvoiceService invoiceService;
+  private final InvoiceLineService invoiceLineService;
+
+  public CompositeOrderTotalFieldsPopulateService(TransactionService transactionService,
+                                                  InvoiceService invoiceService,
+                                                  InvoiceLineService invoiceLineService) {
+    this.transactionService = transactionService;
+    this.invoiceService = invoiceService;
+    this.invoiceLineService = invoiceLineService;
+  }
+
+  @Override
+  public Future<CompositeOrderRetrieveHolder> populate(CompositeOrderRetrieveHolder holder, RequestContext requestContext) {
+    if (holder.getFiscalYear() == null) {
+      return Future.succeededFuture(holder.withTotalExpended(0d).withTotalCredited(0d).withTotalEncumbered(0d));
+    }
+    // We need to fetch invoices, invoice lines and transactions in order to calculate total fields
+    // totalEncumbered = sum of transactions amounts
+    // totalExpended = sum of invoice lines positive total values
+    // totalCredited = absolute sum of invoice lines negative total values
+    var query = "transactionType==Encumbrance AND encumbrance.sourcePurchaseOrderId==%s AND fiscalYearId==%s"
+      .formatted(holder.getOrderId(), holder.getFiscalYearId());
+    return invoiceService.getInvoicesByOrderId(holder.getOrderId(), requestContext)
+      .compose(invoices -> getInvoiceLinesByInvoiceIds(invoices, holder.getFiscalYear(), requestContext)
+        .map(invoiceLines -> groupInvoiceLinesByInvoices(invoices, invoiceLines))
+        .compose(invoiceToInvoiceLinesMap -> transactionService.getTransactions(query, requestContext)
+          .map(transactions -> populateTotalFields(holder, invoiceToInvoiceLinesMap, transactions))))
+      .onFailure(t -> log.error("Failed to populate order's '{}' total fields with fiscal year '{}' from invoice lines",
+        holder.getOrderId(), holder.getFiscalYearId(), t));
+  }
+
+  private Future<List<InvoiceLine>> getInvoiceLinesByInvoiceIds(List<Invoice> invoices, FiscalYear fiscalYear, RequestContext requestContext) {
+    return collectResultsOnSuccess(invoices.stream()
+      .filter(invoice -> invoice.getFiscalYearId().equals(fiscalYear.getId()))
```

Comment:
we had some migration script to populate fiscalYearId for all invoices, but in any case we can use StringUtils.equals to avoid NPE